File: ai-solutions/03/opus-4.5/python/main.py
import sys
from pathlib import Path


def max_subsequence_value(digits: str, keep: int) -> int:
    """
    Return the maximum-valued subsequence of `keep` digits, preserving order.
    Uses a monotonic stack to build the lexicographically largest sequence.
    """
    digits = digits.strip()
    stack: list[str] = []
    n = len(digits)

    for idx, ch in enumerate(digits):
        remaining = n - idx - 1
        while stack and len(stack) + remaining >= keep and stack[-1] < ch:
            stack.pop()
        if len(stack) < keep:
            stack.append(ch)

    return int("".join(stack))


def solve(lines: list[str]) -> tuple[int, int]:
    part1 = 0
    part2 = 0

    for line in lines:
        stripped = line.strip()
        if not stripped:
            continue
        part1 += max_subsequence_value(stripped, 2)
        part2 += max_subsequence_value(stripped, 12)

    return part1, part2


def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: python main.py <input_file>", file=sys.stderr)
        sys.exit(1)

    input_path = Path(sys.argv[1])
    if not input_path.exists():
        print(f"Input file not found: {input_path}", file=sys.stderr)
        sys.exit(1)

    lines = input_path.read_text().splitlines()
    part1, part2 = solve(lines)

    print(part1)
    print(part2)


if __name__ == "__main__":
    main()

File: ai-solutions/03/opus-4.5/rust/src/main.rs
use std::env;
use std::fs;
use std::process;

fn max_subsequence_value(digits: &str, keep: usize) -> u128 {
    let bytes = digits.as_bytes();
    let n = bytes.len();
    let mut stack: Vec<u8> = Vec::with_capacity(keep);

    for (idx, &b) in bytes.iter().enumerate() {
        let remaining = n - idx - 1;
        while !stack.is_empty()
            && stack.len() + remaining >= keep
            && stack.last().copied().unwrap() < b
        {
            stack.pop();
        }
        if stack.len() < keep {
            stack.push(b);
        }
    }

    let value_str: String = stack.into_iter().map(|b| b as char).collect();
    value_str.parse::<u128>().unwrap()
}

fn solve(input: &str) -> (u128, u128) {
    let mut part1: u128 = 0;
    let mut part2: u128 = 0;

    for line in input.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        part1 += max_subsequence_value(trimmed, 2);
        part2 += max_subsequence_value(trimmed, 12);
    }

    (part1, part2)
}

fn main() {
    let mut args = env::args().skip(1);
    let Some(input_path) = args.next() else {
        eprintln!("Usage: cargo run -- <input_file>");
        process::exit(1);
    };

    let contents = fs::read_to_string(&input_path).unwrap_or_else(|err| {
        eprintln!("Failed to read input file {input_path}: {err}");
        process::exit(1);
    });

    let (part1, part2) = solve(&contents);
    println!("{part1}");
    println!("{part2}");
}
